// @blocknote/shadcn/dist/blocknote-shadcn.js
import * as p from "react";
import ze, { createContext as lt, useContext as dt, forwardRef as C, useRef as We, useEffect as Ye, useMemo as Ue } from "react";
import { assertEmpty as y, mergeCSSClasses as ct } from "@blocknote/core";
import { elementOverflow as Ge, mergeRefs as Ke, ComponentsContext as ut, BlockNoteViewRaw as bt } from "@blocknote/react";
import { cva as Q } from "class-variance-authority";
import { clsx as ft } from "clsx";
import { extendTailwindMerge as mt } from "tailwind-merge";
import { Slot as qe } from "@radix-ui/react-slot";
import * as T from "@radix-ui/react-dropdown-menu";
import { ChevronRight as Je, Check as He, Circle as pt, ChevronDown as Xe, ChevronUp as gt } from "lucide-react";
import { FormProvider as vt, useFormContext as xt, useForm as ht } from "react-hook-form";
import * as Ze from "@radix-ui/react-label";
import * as ee from "@radix-ui/react-popover";
import * as D from "@radix-ui/react-select";
import * as V from "@radix-ui/react-tabs";
import * as Qe from "@radix-ui/react-toggle";
import * as q from "@radix-ui/react-tooltip";
var ue = { exports: {} }, G = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Be;
function yt() {
  if (Be) return G;
  Be = 1;
  var r = ze, t = Symbol.for("react.element"), n = Symbol.for("react.fragment"), o = Object.prototype.hasOwnProperty, s = r.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, d = { key: !0, ref: !0, __self: !0, __source: !0 };
  function c(v, f, u) {
    var g, j = {}, _ = null, F = null;
    u !== void 0 && (_ = "" + u), f.key !== void 0 && (_ = "" + f.key), f.ref !== void 0 && (F = f.ref);
    for (g in f) o.call(f, g) && !d.hasOwnProperty(g) && (j[g] = f[g]);
    if (v && v.defaultProps) for (g in f = v.defaultProps, f) j[g] === void 0 && (j[g] = f[g]);
    return { $$typeof: t, type: v, key: _, ref: F, props: j, _owner: s.current };
  }
  return G.Fragment = n, G.jsx = c, G.jsxs = c, G;
}
var K = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ve;
function wt() {
  return Ve || (Ve = 1, process.env.NODE_ENV !== "production" && function() {
    var r = ze, t = Symbol.for("react.element"), n = Symbol.for("react.portal"), o = Symbol.for("react.fragment"), s = Symbol.for("react.strict_mode"), d = Symbol.for("react.profiler"), c = Symbol.for("react.provider"), v = Symbol.for("react.context"), f = Symbol.for("react.forward_ref"), u = Symbol.for("react.suspense"), g = Symbol.for("react.suspense_list"), j = Symbol.for("react.memo"), _ = Symbol.for("react.lazy"), F = Symbol.for("react.offscreen"), fe = Symbol.iterator, Sn = "@@iterator";
    function Tn(e) {
      if (e === null || typeof e != "object")
        return null;
      var i = fe && e[fe] || e[Sn];
      return typeof i == "function" ? i : null;
    }
    var $ = r.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function P(e) {
      {
        for (var i = arguments.length, l = new Array(i > 1 ? i - 1 : 0), m = 1; m < i; m++)
          l[m - 1] = arguments[m];
        Rn("error", e, l);
      }
    }
    function Rn(e, i, l) {
      {
        var m = $.ReactDebugCurrentFrame, w = m.getStackAddendum();
        w !== "" && (i += "%s", l = l.concat([w]));
        var N = l.map(function(h) {
          return String(h);
        });
        N.unshift("Warning: " + i), Function.prototype.apply.call(console[e], console, N);
      }
    }
    var En = !1, Dn = !1, _n = !1, Pn = !1, In = !1, me;
    me = Symbol.for("react.module.reference");
    function Mn(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === o || e === d || In || e === s || e === u || e === g || Pn || e === F || En || Dn || _n || typeof e == "object" && e !== null && (e.$$typeof === _ || e.$$typeof === j || e.$$typeof === c || e.$$typeof === v || e.$$typeof === f || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === me || e.getModuleId !== void 0));
    }
    function kn(e, i, l) {
      var m = e.displayName;
      if (m)
        return m;
      var w = i.displayName || i.name || "";
      return w !== "" ? l + "(" + w + ")" : l;
    }
    function pe(e) {
      return e.displayName || "Context";
    }
    function O(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && P("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case o:
          return "Fragment";
        case n:
          return "Portal";
        case d:
          return "Profiler";
        case s:
          return "StrictMode";
        case u:
          return "Suspense";
        case g:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case v:
            var i = e;
            return pe(i) + ".Consumer";
          case c:
            var l = e;
            return pe(l._context) + ".Provider";
          case f:
            return kn(e, e.render, "ForwardRef");
          case j:
            var m = e.displayName || null;
            return m !== null ? m : O(e.type) || "Memo";
          case _: {
            var w = e, N = w._payload, h = w._init;
            try {
              return O(h(N));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var L = Object.assign, W = 0, ge, ve, xe, he, ye, we, Ce;
    function Ne() {
    }
    Ne.__reactDisabledLog = !0;
    function Fn() {
      {
        if (W === 0) {
          ge = console.log, ve = console.info, xe = console.warn, he = console.error, ye = console.group, we = console.groupCollapsed, Ce = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: Ne,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        W++;
      }
    }
    function On() {
      {
        if (W--, W === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: L({}, e, {
              value: ge
            }),
            info: L({}, e, {
              value: ve
            }),
            warn: L({}, e, {
              value: xe
            }),
            error: L({}, e, {
              value: he
            }),
            group: L({}, e, {
              value: ye
            }),
            groupCollapsed: L({}, e, {
              value: we
            }),
            groupEnd: L({}, e, {
              value: Ce
            })
          });
        }
        W < 0 && P("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var te = $.ReactCurrentDispatcher, re;
    function J(e, i, l) {
      {
        if (re === void 0)
          try {
            throw Error();
          } catch (w) {
            var m = w.stack.trim().match(/\n( *(at )?)/);
            re = m && m[1] || "";
          }
        return `
` + re + e;
      }
    }
    var oe = !1, H;
    {
      var Ln = typeof WeakMap == "function" ? WeakMap : Map;
      H = new Ln();
    }
    function je(e, i) {
      if (!e || oe)
        return "";
      {
        var l = H.get(e);
        if (l !== void 0)
          return l;
      }
      var m;
      oe = !0;
      var w = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var N;
      N = te.current, te.current = null, Fn();
      try {
        if (i) {
          var h = function() {
            throw Error();
          };
          if (Object.defineProperty(h.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(h, []);
            } catch (M) {
              m = M;
            }
            Reflect.construct(e, [], h);
          } else {
            try {
              h.call();
            } catch (M) {
              m = M;
            }
            e.call(h.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (M) {
            m = M;
          }
          e();
        }
      } catch (M) {
        if (M && m && typeof M.stack == "string") {
          for (var x = M.stack.split(`
`), I = m.stack.split(`
`), S = x.length - 1, E = I.length - 1; S >= 1 && E >= 0 && x[S] !== I[E]; )
            E--;
          for (; S >= 1 && E >= 0; S--, E--)
            if (x[S] !== I[E]) {
              if (S !== 1 || E !== 1)
                do
                  if (S--, E--, E < 0 || x[S] !== I[E]) {
                    var k = `
` + x[S].replace(" at new ", " at ");
                    return e.displayName && k.includes("<anonymous>") && (k = k.replace("<anonymous>", e.displayName)), typeof e == "function" && H.set(e, k), k;
                  }
                while (S >= 1 && E >= 0);
              break;
            }
        }
      } finally {
        oe = !1, te.current = N, On(), Error.prepareStackTrace = w;
      }
      var z = e ? e.displayName || e.name : "", B = z ? J(z) : "";
      return typeof e == "function" && H.set(e, B), B;
    }
    function Bn(e, i, l) {
      return je(e, !1);
    }
    function Vn(e) {
      var i = e.prototype;
      return !!(i && i.isReactComponent);
    }
    function X(e, i, l) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return je(e, Vn(e));
      if (typeof e == "string")
        return J(e);
      switch (e) {
        case u:
          return J("Suspense");
        case g:
          return J("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case f:
            return Bn(e.render);
          case j:
            return X(e.type, i, l);
          case _: {
            var m = e, w = m._payload, N = m._init;
            try {
              return X(N(w), i, l);
            } catch {
            }
          }
        }
      return "";
    }
    var Y = Object.prototype.hasOwnProperty, Se = {}, Te = $.ReactDebugCurrentFrame;
    function Z(e) {
      if (e) {
        var i = e._owner, l = X(e.type, e._source, i ? i.type : null);
        Te.setExtraStackFrame(l);
      } else
        Te.setExtraStackFrame(null);
    }
    function $n(e, i, l, m, w) {
      {
        var N = Function.call.bind(Y);
        for (var h in e)
          if (N(e, h)) {
            var x = void 0;
            try {
              if (typeof e[h] != "function") {
                var I = Error((m || "React class") + ": " + l + " type `" + h + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[h] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw I.name = "Invariant Violation", I;
              }
              x = e[h](i, h, m, l, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (S) {
              x = S;
            }
            x && !(x instanceof Error) && (Z(w), P("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", m || "React class", l, h, typeof x), Z(null)), x instanceof Error && !(x.message in Se) && (Se[x.message] = !0, Z(w), P("Failed %s type: %s", l, x.message), Z(null));
          }
      }
    }
    var An = Array.isArray;
    function ae(e) {
      return An(e);
    }
    function zn(e) {
      {
        var i = typeof Symbol == "function" && Symbol.toStringTag, l = i && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return l;
      }
    }
    function Wn(e) {
      try {
        return Re(e), !1;
      } catch {
        return !0;
      }
    }
    function Re(e) {
      return "" + e;
    }
    function Ee(e) {
      if (Wn(e))
        return P("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", zn(e)), Re(e);
    }
    var U = $.ReactCurrentOwner, Yn = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, De, _e, se;
    se = {};
    function Un(e) {
      if (Y.call(e, "ref")) {
        var i = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (i && i.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function Gn(e) {
      if (Y.call(e, "key")) {
        var i = Object.getOwnPropertyDescriptor(e, "key").get;
        if (i && i.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function Kn(e, i) {
      if (typeof e.ref == "string" && U.current && i && U.current.stateNode !== i) {
        var l = O(U.current.type);
        se[l] || (P('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', O(U.current.type), e.ref), se[l] = !0);
      }
    }
    function qn(e, i) {
      {
        var l = function() {
          De || (De = !0, P("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", i));
        };
        l.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: l,
          configurable: !0
        });
      }
    }
    function Jn(e, i) {
      {
        var l = function() {
          _e || (_e = !0, P("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", i));
        };
        l.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: l,
          configurable: !0
        });
      }
    }
    var Hn = function(e, i, l, m, w, N, h) {
      var x = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: t,
        // Built-in properties that belong on the element
        type: e,
        key: i,
        ref: l,
        props: h,
        // Record the component responsible for creating this element.
        _owner: N
      };
      return x._store = {}, Object.defineProperty(x._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(x, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: m
      }), Object.defineProperty(x, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: w
      }), Object.freeze && (Object.freeze(x.props), Object.freeze(x)), x;
    };
    function Xn(e, i, l, m, w) {
      {
        var N, h = {}, x = null, I = null;
        l !== void 0 && (Ee(l), x = "" + l), Gn(i) && (Ee(i.key), x = "" + i.key), Un(i) && (I = i.ref, Kn(i, w));
        for (N in i)
          Y.call(i, N) && !Yn.hasOwnProperty(N) && (h[N] = i[N]);
        if (e && e.defaultProps) {
          var S = e.defaultProps;
          for (N in S)
            h[N] === void 0 && (h[N] = S[N]);
        }
        if (x || I) {
          var E = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          x && qn(h, E), I && Jn(h, E);
        }
        return Hn(e, x, I, w, m, U.current, h);
      }
    }
    var ie = $.ReactCurrentOwner, Pe = $.ReactDebugCurrentFrame;
    function A(e) {
      if (e) {
        var i = e._owner, l = X(e.type, e._source, i ? i.type : null);
        Pe.setExtraStackFrame(l);
      } else
        Pe.setExtraStackFrame(null);
    }
    var le;
    le = !1;
    function de(e) {
      return typeof e == "object" && e !== null && e.$$typeof === t;
    }
    function Ie() {
      {
        if (ie.current) {
          var e = O(ie.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function Zn(e) {
      return "";
    }
    var Me = {};
    function Qn(e) {
      {
        var i = Ie();
        if (!i) {
          var l = typeof e == "string" ? e : e.displayName || e.name;
          l && (i = `

Check the top-level render call using <` + l + ">.");
        }
        return i;
      }
    }
    function ke(e, i) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var l = Qn(i);
        if (Me[l])
          return;
        Me[l] = !0;
        var m = "";
        e && e._owner && e._owner !== ie.current && (m = " It was passed a child from " + O(e._owner.type) + "."), A(e), P('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', l, m), A(null);
      }
    }
    function Fe(e, i) {
      {
        if (typeof e != "object")
          return;
        if (ae(e))
          for (var l = 0; l < e.length; l++) {
            var m = e[l];
            de(m) && ke(m, i);
          }
        else if (de(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var w = Tn(e);
          if (typeof w == "function" && w !== e.entries)
            for (var N = w.call(e), h; !(h = N.next()).done; )
              de(h.value) && ke(h.value, i);
        }
      }
    }
    function et(e) {
      {
        var i = e.type;
        if (i == null || typeof i == "string")
          return;
        var l;
        if (typeof i == "function")
          l = i.propTypes;
        else if (typeof i == "object" && (i.$$typeof === f || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        i.$$typeof === j))
          l = i.propTypes;
        else
          return;
        if (l) {
          var m = O(i);
          $n(l, e.props, "prop", m, e);
        } else if (i.PropTypes !== void 0 && !le) {
          le = !0;
          var w = O(i);
          P("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", w || "Unknown");
        }
        typeof i.getDefaultProps == "function" && !i.getDefaultProps.isReactClassApproved && P("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function nt(e) {
      {
        for (var i = Object.keys(e.props), l = 0; l < i.length; l++) {
          var m = i[l];
          if (m !== "children" && m !== "key") {
            A(e), P("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", m), A(null);
            break;
          }
        }
        e.ref !== null && (A(e), P("Invalid attribute `ref` supplied to `React.Fragment`."), A(null));
      }
    }
    var Oe = {};
    function Le(e, i, l, m, w, N) {
      {
        var h = Mn(e);
        if (!h) {
          var x = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (x += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var I = Zn();
          I ? x += I : x += Ie();
          var S;
          e === null ? S = "null" : ae(e) ? S = "array" : e !== void 0 && e.$$typeof === t ? (S = "<" + (O(e.type) || "Unknown") + " />", x = " Did you accidentally export a JSX literal instead of a component?") : S = typeof e, P("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", S, x);
        }
        var E = Xn(e, i, l, w, N);
        if (E == null)
          return E;
        if (h) {
          var k = i.children;
          if (k !== void 0)
            if (m)
              if (ae(k)) {
                for (var z = 0; z < k.length; z++)
                  Fe(k[z], e);
                Object.freeze && Object.freeze(k);
              } else
                P("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Fe(k, e);
        }
        if (Y.call(i, "key")) {
          var B = O(e), M = Object.keys(i).filter(function(it) {
            return it !== "key";
          }), ce = M.length > 0 ? "{key: someKey, " + M.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!Oe[B + ce]) {
            var st = M.length > 0 ? "{" + M.join(": ..., ") + ": ...}" : "{}";
            P(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, ce, B, st, B), Oe[B + ce] = !0;
          }
        }
        return e === o ? nt(E) : et(E), E;
      }
    }
    function tt(e, i, l) {
      return Le(e, i, l, !0);
    }
    function rt(e, i, l) {
      return Le(e, i, l, !1);
    }
    var ot = rt, at = tt;
    K.Fragment = o, K.jsx = ot, K.jsxs = at;
  }()), K;
}
process.env.NODE_ENV === "production" ? ue.exports = yt() : ue.exports = wt();
var a = ue.exports;
const Ct = mt({
  prefix: "bn-"
});
function b(...r) {
  return Ct(ft(r));
}
const Nt = Q(
  "bn-inline-flex bn-items-center bn-rounded-full bn-border bn-px-2.5 bn-py-0.5 bn-text-xs bn-font-semibold bn-transition-colors focus:bn-outline-none focus:bn-ring-2 focus:bn-ring-ring focus:bn-ring-offset-2",
  {
    variants: {
      variant: {
        default: "bn-border-transparent bn-bg-primary bn-text-primary-foreground hover:bn-bg-primary/80",
        secondary: "bn-border-transparent bn-bg-secondary bn-text-secondary-foreground hover:bn-bg-secondary/80",
        destructive: "bn-border-transparent bn-bg-destructive bn-text-destructive-foreground hover:bn-bg-destructive/80",
        outline: "bn-text-foreground"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function jt({ className: r, variant: t, ...n }) {
  return /* @__PURE__ */ a.jsx("div", { className: b(Nt({ variant: t }), r), ...n });
}
const St = Q(
  "bn-inline-flex bn-items-center bn-justify-center bn-whitespace-nowrap bn-rounded-md bn-text-sm bn-font-medium bn-ring-offset-background bn-transition-colors focus-visible:bn-outline-none focus-visible:bn-ring-2 focus-visible:bn-ring-ring focus-visible:bn-ring-offset-2 disabled:bn-pointer-events-none disabled:bn-opacity-50",
  {
    variants: {
      variant: {
        default: "bn-bg-primary bn-text-primary-foreground hover:bn-bg-primary/90",
        destructive: "bn-bg-destructive bn-text-destructive-foreground hover:bn-bg-destructive/90",
        outline: "bn-border bn-border-input bn-bg-background hover:bn-bg-accent hover:bn-text-accent-foreground",
        secondary: "bn-bg-secondary bn-text-secondary-foreground hover:bn-bg-secondary/80",
        ghost: "hover:bn-bg-accent hover:bn-text-accent-foreground",
        link: "bn-text-primary bn-underline-offset-4 hover:bn-underline"
      },
      size: {
        default: "bn-h-10 bn-px-4 bn-py-2",
        sm: "bn-h-9 bn-rounded-md bn-px-3",
        lg: "bn-h-11 bn-rounded-md bn-px-8",
        icon: "bn-h-10 bn-w-10"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
), en = p.forwardRef(
  ({ className: r, variant: t, size: n, asChild: o = !1, ...s }, d) => {
    const c = o ? qe : "button";
    return /* @__PURE__ */ a.jsx(
      c,
      {
        className: b(St({ variant: t, size: n, className: r })),
        ref: d,
        ...s
      }
    );
  }
);
en.displayName = "Button";
const nn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  "div",
  {
    ref: n,
    className: b(
      "bn-rounded-lg bn-border bn-bg-card bn-text-card-foreground bn-shadow-sm",
      r
    ),
    ...t
  }
));
nn.displayName = "Card";
const Tt = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  "div",
  {
    ref: n,
    className: b("bn-flex bn-flex-col bn-space-y-1.5 bn-p-6", r),
    ...t
  }
));
Tt.displayName = "CardHeader";
const Rt = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  "h3",
  {
    ref: n,
    className: b(
      "bn-text-2xl bn-font-semibold bn-leading-none bn-tracking-tight",
      r
    ),
    ...t
  }
));
Rt.displayName = "CardTitle";
const Et = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  "p",
  {
    ref: n,
    className: b("bn-text-sm bn-text-muted-foreground", r),
    ...t
  }
));
Et.displayName = "CardDescription";
const tn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx("div", { ref: n, className: b("bn-p-6 bn-pt-0", r), ...t }));
tn.displayName = "CardContent";
const Dt = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  "div",
  {
    ref: n,
    className: b("bn-flex bn-items-center bn-p-6 bn-pt-0", r),
    ...t
  }
));
Dt.displayName = "CardFooter";
const _t = T.Root, Pt = T.Trigger, It = T.Sub, rn = p.forwardRef(({ className: r, inset: t, children: n, ...o }, s) => /* @__PURE__ */ a.jsxs(
  T.SubTrigger,
  {
    ref: s,
    className: b(
      "bn-flex bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-px-2 bn-py-1.5 bn-text-sm bn-outline-none focus:bn-bg-accent data-[state=open]:bn-bg-accent",
      t && "bn-pl-8",
      r
    ),
    ...o,
    children: [
      n,
      /* @__PURE__ */ a.jsx(Je, { className: "bn-ml-auto bn-h-4 bn-w-4" })
    ]
  }
));
rn.displayName = T.SubTrigger.displayName;
const on = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  T.SubContent,
  {
    ref: n,
    className: b(
      "bn-z-50 bn-min-w-[8rem] bn-overflow-hidden bn-rounded-md bn-border bn-bg-popover bn-p-1 bn-text-popover-foreground bn-shadow-lg data-[state=open]:bn-animate-in data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=open]:bn-fade-in-0 data-[state=closed]:bn-zoom-out-95 data-[state=open]:bn-zoom-in-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
      r
    ),
    ...t
  }
));
on.displayName = T.SubContent.displayName;
const an = p.forwardRef(({ className: r, sideOffset: t = 4, ...n }, o) => (
  // <DropdownMenuPrimitive.Portal>
  /* @__PURE__ */ a.jsx(
    T.Content,
    {
      ref: o,
      sideOffset: t,
      className: b(
        "bn-z-50 bn-min-w-[8rem] bn-overflow-hidden bn-rounded-md bn-border bn-bg-popover bn-p-1 bn-text-popover-foreground bn-shadow-md data-[state=open]:bn-animate-in data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=open]:bn-fade-in-0 data-[state=closed]:bn-zoom-out-95 data-[state=open]:bn-zoom-in-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
        r
      ),
      ...n
    }
  )
));
an.displayName = T.Content.displayName;
const sn = p.forwardRef(({ className: r, inset: t, ...n }, o) => /* @__PURE__ */ a.jsx(
  T.Item,
  {
    ref: o,
    className: b(
      "bn-relative bn-flex bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-px-2 bn-py-1.5 bn-text-sm bn-outline-none bn-transition-colors focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
      t && "bn-pl-8",
      r
    ),
    ...n
  }
));
sn.displayName = T.Item.displayName;
const ln = p.forwardRef(({ className: r, children: t, checked: n, ...o }, s) => /* @__PURE__ */ a.jsxs(
  T.CheckboxItem,
  {
    ref: s,
    className: b(
      "bn-relative bn-flex bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-py-1.5 bn-pl-8 bn-pr-2 bn-text-sm bn-outline-none bn-transition-colors focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
      r
    ),
    checked: n,
    ...o,
    children: [
      /* @__PURE__ */ a.jsx("span", { className: "bn-absolute bn-left-2 bn-flex bn-h-3.5 bn-w-3.5 bn-items-center bn-justify-center", children: /* @__PURE__ */ a.jsx(T.ItemIndicator, { children: /* @__PURE__ */ a.jsx(He, { className: "bn-h-4 bn-w-4" }) }) }),
      t
    ]
  }
));
ln.displayName = T.CheckboxItem.displayName;
const Mt = p.forwardRef(({ className: r, children: t, ...n }, o) => /* @__PURE__ */ a.jsxs(
  T.RadioItem,
  {
    ref: o,
    className: b(
      "bn-relative bn-flex bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-py-1.5 bn-pl-8 bn-pr-2 bn-text-sm bn-outline-none bn-transition-colors focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
      r
    ),
    ...n,
    children: [
      /* @__PURE__ */ a.jsx("span", { className: "bn-absolute bn-left-2 bn-flex bn-h-3.5 bn-w-3.5 bn-items-center bn-justify-center", children: /* @__PURE__ */ a.jsx(T.ItemIndicator, { children: /* @__PURE__ */ a.jsx(pt, { className: "bn-h-2 bn-w-2 bn-fill-current" }) }) }),
      t
    ]
  }
));
Mt.displayName = T.RadioItem.displayName;
const dn = p.forwardRef(({ className: r, inset: t, ...n }, o) => /* @__PURE__ */ a.jsx(
  T.Label,
  {
    ref: o,
    className: b(
      "bn-px-2 bn-py-1.5 bn-text-sm bn-font-semibold",
      t && "bn-pl-8",
      r
    ),
    ...n
  }
));
dn.displayName = T.Label.displayName;
const cn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  T.Separator,
  {
    ref: n,
    className: b("bn--mx-1 bn-my-1 bn-h-px bn-bg-muted", r),
    ...t
  }
));
cn.displayName = T.Separator.displayName;
const kt = Q(
  "bn-text-sm bn-font-medium bn-leading-none peer-disabled:bn-cursor-not-allowed peer-disabled:bn-opacity-70"
), be = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  Ze.Root,
  {
    ref: n,
    className: b(kt(), r),
    ...t
  }
));
be.displayName = Ze.Root.displayName;
const Ft = vt, Ot = p.createContext(
  {}
), ne = () => {
  const r = p.useContext(Ot), t = p.useContext(un), { getFieldState: n, formState: o } = xt(), s = n(r.name, o);
  if (!r)
    throw new Error("useFormField should be used within <FormField>");
  const { id: d } = t;
  return {
    id: d,
    name: r.name,
    formItemId: `${d}-form-item`,
    formDescriptionId: `${d}-form-item-description`,
    formMessageId: `${d}-form-item-message`,
    ...s
  };
}, un = p.createContext(
  {}
), Lt = p.forwardRef(({ className: r, ...t }, n) => {
  const o = p.useId();
  return /* @__PURE__ */ a.jsx(un.Provider, { value: { id: o }, children: /* @__PURE__ */ a.jsx("div", { ref: n, className: b("bn-space-y-2", r), ...t }) });
});
Lt.displayName = "FormItem";
const Bt = p.forwardRef(({ className: r, ...t }, n) => {
  const { error: o, formItemId: s } = ne();
  return /* @__PURE__ */ a.jsx(
    be,
    {
      ref: n,
      className: b(o && "bn-text-destructive", r),
      htmlFor: s,
      ...t
    }
  );
});
Bt.displayName = "FormLabel";
const Vt = p.forwardRef(({ ...r }, t) => {
  const { error: n, formItemId: o, formDescriptionId: s, formMessageId: d } = ne();
  return /* @__PURE__ */ a.jsx(
    qe,
    {
      ref: t,
      id: o,
      "aria-describedby": n ? `${s} ${d}` : `${s}`,
      "aria-invalid": !!n,
      ...r
    }
  );
});
Vt.displayName = "FormControl";
const $t = p.forwardRef(({ className: r, ...t }, n) => {
  const { formDescriptionId: o } = ne();
  return /* @__PURE__ */ a.jsx(
    "p",
    {
      ref: n,
      id: o,
      className: b("bn-text-sm bn-text-muted-foreground", r),
      ...t
    }
  );
});
$t.displayName = "FormDescription";
const At = p.forwardRef(({ className: r, children: t, ...n }, o) => {
  const { error: s, formMessageId: d } = ne(), c = s ? String(s == null ? void 0 : s.message) : t;
  return c ? /* @__PURE__ */ a.jsx(
    "p",
    {
      ref: o,
      id: d,
      className: b("bn-text-sm bn-font-medium bn-text-destructive", r),
      ...n,
      children: c
    }
  ) : null;
});
At.displayName = "FormMessage";
const bn = p.forwardRef(
  ({ className: r, type: t, ...n }, o) => /* @__PURE__ */ a.jsx(
    "input",
    {
      type: t,
      className: b(
        "bn-flex bn-h-10 bn-w-full bn-rounded-md bn-border bn-border-input bn-bg-background bn-px-3 bn-py-2 bn-text-sm bn-ring-offset-background file:bn-border-0 file:bn-bg-transparent file:bn-text-sm file:bn-font-medium placeholder:bn-text-muted-foreground focus-visible:bn-outline-none focus-visible:bn-ring-2 focus-visible:bn-ring-ring focus-visible:bn-ring-offset-2 disabled:bn-cursor-not-allowed disabled:bn-opacity-50",
        r
      ),
      ref: o,
      ...n
    }
  )
);
bn.displayName = "Input";
const zt = ee.Root, Wt = ee.Trigger, fn = p.forwardRef(({ className: r, align: t = "center", sideOffset: n = 4, ...o }, s) => (
  // <PopoverPrimitive.Portal>
  /* @__PURE__ */ a.jsx(
    ee.Content,
    {
      ref: s,
      align: t,
      sideOffset: n,
      className: b(
        "bn-z-50 bn-w-72 bn-rounded-md bn-border bn-bg-popover bn-p-4 bn-text-popover-foreground bn-shadow-md bn-outline-none data-[state=open]:bn-animate-in data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=open]:bn-fade-in-0 data-[state=closed]:bn-zoom-out-95 data-[state=open]:bn-zoom-in-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
        r
      ),
      ...o
    }
  )
));
fn.displayName = ee.Content.displayName;
const Yt = D.Root, Ut = D.Value, mn = p.forwardRef(({ className: r, children: t, ...n }, o) => /* @__PURE__ */ a.jsxs(
  D.Trigger,
  {
    ref: o,
    className: b(
      "bn-flex bn-h-10 bn-w-full bn-items-center bn-justify-between bn-rounded-md bn-border bn-border-input bn-bg-background bn-px-3 bn-py-2 bn-text-sm bn-ring-offset-background placeholder:bn-text-muted-foreground focus:bn-outline-none focus:bn-ring-2 focus:bn-ring-ring focus:bn-ring-offset-2 disabled:bn-cursor-not-allowed disabled:bn-opacity-50 [&>span]:bn-line-clamp-1",
      r
    ),
    ...n,
    children: [
      t,
      /* @__PURE__ */ a.jsx(D.Icon, { asChild: !0, children: /* @__PURE__ */ a.jsx(Xe, { className: "bn-h-4 bn-w-4 bn-opacity-50" }) })
    ]
  }
));
mn.displayName = D.Trigger.displayName;
const pn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  D.ScrollUpButton,
  {
    ref: n,
    className: b(
      "bn-flex bn-cursor-default bn-items-center bn-justify-center bn-py-1",
      r
    ),
    ...t,
    children: /* @__PURE__ */ a.jsx(gt, { className: "bn-h-4 bn-w-4" })
  }
));
pn.displayName = D.ScrollUpButton.displayName;
const gn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  D.ScrollDownButton,
  {
    ref: n,
    className: b(
      "bn-flex bn-cursor-default bn-items-center bn-justify-center bn-py-1",
      r
    ),
    ...t,
    children: /* @__PURE__ */ a.jsx(Xe, { className: "bn-h-4 bn-w-4" })
  }
));
gn.displayName = D.ScrollDownButton.displayName;
const vn = p.forwardRef(({ className: r, children: t, position: n = "popper", ...o }, s) => (
  // <SelectPrimitive.Portal>
  /* @__PURE__ */ a.jsxs(
    D.Content,
    {
      ref: s,
      className: b(
        "bn-relative bn-z-50 bn-max-h-96 bn-min-w-[8rem] bn-overflow-hidden bn-rounded-md bn-border bn-bg-popover bn-text-popover-foreground bn-shadow-md data-[state=open]:bn-animate-in data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=open]:bn-fade-in-0 data-[state=closed]:bn-zoom-out-95 data-[state=open]:bn-zoom-in-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
        n === "popper" && "data-[side=bottom]:bn-translate-y-1 data-[side=left]:bn--translate-x-1 data-[side=right]:bn-translate-x-1 data-[side=top]:bn--translate-y-1",
        r
      ),
      position: n,
      ...o,
      children: [
        /* @__PURE__ */ a.jsx(pn, {}),
        /* @__PURE__ */ a.jsx(
          D.Viewport,
          {
            className: b(
              "bn-p-1",
              n === "popper" && "bn-h-[var(--radix-select-trigger-height)] bn-w-full bn-min-w-[var(--radix-select-trigger-width)]"
            ),
            children: t
          }
        ),
        /* @__PURE__ */ a.jsx(gn, {})
      ]
    }
  )
));
vn.displayName = D.Content.displayName;
const Gt = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  D.Label,
  {
    ref: n,
    className: b(
      "bn-py-1.5 bn-pl-8 bn-pr-2 bn-text-sm bn-font-semibold",
      r
    ),
    ...t
  }
));
Gt.displayName = D.Label.displayName;
const xn = p.forwardRef(({ className: r, children: t, ...n }, o) => /* @__PURE__ */ a.jsxs(
  D.Item,
  {
    ref: o,
    className: b(
      "bn-relative bn-flex bn-w-full bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-py-1.5 bn-pl-8 bn-pr-2 bn-text-sm bn-outline-none focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
      r
    ),
    ...n,
    children: [
      /* @__PURE__ */ a.jsx("span", { className: "bn-absolute bn-left-2 bn-flex bn-h-3.5 bn-w-3.5 bn-items-center bn-justify-center", children: /* @__PURE__ */ a.jsx(D.ItemIndicator, { children: /* @__PURE__ */ a.jsx(He, { className: "bn-h-4 bn-w-4" }) }) }),
      /* @__PURE__ */ a.jsx(D.ItemText, { children: t })
    ]
  }
));
xn.displayName = D.Item.displayName;
const Kt = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  D.Separator,
  {
    ref: n,
    className: b("bn--mx-1 bn-my-1 bn-h-px bn-bg-muted", r),
    ...t
  }
));
Kt.displayName = D.Separator.displayName;
const qt = V.Root, hn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  V.List,
  {
    ref: n,
    className: b(
      "bn-inline-flex bn-h-10 bn-items-center bn-justify-center bn-rounded-md bn-bg-muted bn-p-1 bn-text-muted-foreground",
      r
    ),
    ...t
  }
));
hn.displayName = V.List.displayName;
const yn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  V.Trigger,
  {
    ref: n,
    className: b(
      "bn-inline-flex bn-items-center bn-justify-center bn-whitespace-nowrap bn-rounded-sm bn-px-3 bn-py-1.5 bn-text-sm bn-font-medium bn-ring-offset-background bn-transition-all focus-visible:bn-outline-none focus-visible:bn-ring-2 focus-visible:bn-ring-ring focus-visible:bn-ring-offset-2 disabled:bn-pointer-events-none disabled:bn-opacity-50 data-[state=active]:bn-bg-background data-[state=active]:bn-text-foreground data-[state=active]:bn-shadow-sm",
      r
    ),
    ...t
  }
));
yn.displayName = V.Trigger.displayName;
const wn = p.forwardRef(({ className: r, ...t }, n) => /* @__PURE__ */ a.jsx(
  V.Content,
  {
    ref: n,
    className: b(
      "bn-mt-2 bn-ring-offset-background focus-visible:bn-outline-none focus-visible:bn-ring-2 focus-visible:bn-ring-ring focus-visible:bn-ring-offset-2",
      r
    ),
    ...t
  }
));
wn.displayName = V.Content.displayName;
const Jt = Q(
  "bn-inline-flex bn-items-center bn-justify-center bn-rounded-md bn-text-sm bn-font-medium bn-ring-offset-background bn-transition-colors hover:bn-bg-muted hover:bn-text-muted-foreground focus-visible:bn-outline-none focus-visible:bn-ring-2 focus-visible:bn-ring-ring focus-visible:bn-ring-offset-2 disabled:bn-pointer-events-none disabled:bn-opacity-50 data-[state=on]:bn-bg-accent data-[state=on]:bn-text-accent-foreground",
  {
    variants: {
      variant: {
        default: "bn-bg-transparent",
        outline: "bn-border bn-border-input bn-bg-transparent hover:bn-bg-accent hover:bn-text-accent-foreground"
      },
      size: {
        default: "bn-h-10 bn-px-3",
        sm: "bn-h-9 bn-px-2.5",
        lg: "bn-h-11 bn-px-5"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
), Cn = p.forwardRef(({ className: r, variant: t, size: n, ...o }, s) => /* @__PURE__ */ a.jsx(
  Qe.Root,
  {
    ref: s,
    className: b(Jt({ variant: t, size: n, className: r })),
    ...o
  }
));
Cn.displayName = Qe.Root.displayName;
const Ht = q.Provider, Xt = q.Root, Zt = q.Trigger, Nn = p.forwardRef(({ className: r, sideOffset: t = 4, ...n }, o) => /* @__PURE__ */ a.jsx(
  q.Content,
  {
    ref: o,
    sideOffset: t,
    className: b(
      "bn-z-50 bn-overflow-hidden bn-rounded-md bn-border bn-bg-popover bn-px-3 bn-py-1.5 bn-text-sm bn-text-popover-foreground bn-shadow-md bn-animate-in bn-fade-in-0 bn-zoom-in-95 data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=closed]:bn-zoom-out-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
      r
    ),
    ...n
  }
));
Nn.displayName = q.Content.displayName;
const Qt = {
  Badge: {
    Badge: jt
  },
  Button: {
    Button: en
  },
  Card: {
    Card: nn,
    CardContent: tn
  },
  DropdownMenu: {
    DropdownMenu: _t,
    DropdownMenuCheckboxItem: ln,
    DropdownMenuContent: an,
    DropdownMenuItem: sn,
    DropdownMenuLabel: dn,
    DropdownMenuSeparator: cn,
    DropdownMenuSub: It,
    DropdownMenuSubContent: on,
    DropdownMenuSubTrigger: rn,
    DropdownMenuTrigger: Pt
  },
  Form: {
    Form: Ft
  },
  Input: {
    Input: bn
  },
  Label: {
    Label: be
  },
  Popover: {
    Popover: zt,
    PopoverContent: fn,
    PopoverTrigger: Wt
  },
  Select: {
    Select: Yt,
    SelectContent: vn,
    SelectItem: xn,
    SelectTrigger: mn,
    SelectValue: Ut
  },
  Tabs: {
    Tabs: qt,
    TabsContent: wn,
    TabsList: hn,
    TabsTrigger: yn
  },
  Toggle: {
    Toggle: Cn
  },
  Tooltip: {
    Tooltip: Xt,
    TooltipContent: Nn,
    TooltipProvider: Ht,
    TooltipTrigger: Zt
  }
}, jn = lt(void 0);
function R() {
  return dt(jn);
}
const er = (r) => {
  const { children: t, ...n } = r;
  y(n);
  const o = R(), s = ht();
  return /* @__PURE__ */ a.jsx(o.Form.Form, { ...s, children: t });
}, nr = C((r, t) => {
  const { className: n, children: o, id: s, columns: d, ...c } = r;
  return y(c), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: n,
      style: { gridTemplateColumns: `repeat(${d}, 1fr)` },
      ref: t,
      id: s,
      role: "grid",
      children: o
    }
  );
}), tr = C((r, t) => {
  const { className: n, children: o, columns: s, ...d } = r;
  return y(d), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: n,
      style: { gridColumn: `1 / ${s + 1}` },
      ref: t,
      children: o
    }
  );
}), rr = C((r, t) => {
  const { className: n, isSelected: o, onClick: s, item: d, id: c, ...v } = r;
  y(v);
  const f = We(null);
  return Ye(() => {
    if (!f.current || !o)
      return;
    const u = Ge(
      f.current,
      document.querySelector(".bn-grid-suggestion-menu")
    );
    u === "top" ? f.current.scrollIntoView(!0) : u === "bottom" && f.current.scrollIntoView(!1);
  }, [o]), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: n,
      ref: Ke([t, f]),
      id: c,
      role: "option",
      onClick: s,
      "aria-selected": o || void 0,
      children: d.icon
    }
  );
}), or = C((r, t) => {
  const {
    className: n,
    children: o,
    // unused, using "dots" instead
    columns: s,
    ...d
  } = r;
  return y(d), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: n,
      style: { gridColumn: `1 / ${s + 1}` },
      ref: t,
      children: o
    }
  );
}), ar = (r) => C(
  (t, n) => /* @__PURE__ */ a.jsx(
    r,
    {
      onPointerDown: (o) => {
        o.nativeEvent.fakeEvent || (o.ctrlKey = !0);
      },
      onPointerUp: (o) => {
        const s = new PointerEvent("pointerdown", o.nativeEvent);
        s.fakeEvent = !0, o.target.dispatchEvent(s);
      },
      ...t,
      ref: n
    }
  )
), sr = (r) => {
  const {
    children: t,
    onOpenChange: n,
    position: o,
    // Unused
    sub: s,
    ...d
  } = r;
  y(d);
  const c = R();
  return s ? /* @__PURE__ */ a.jsx(
    c.DropdownMenu.DropdownMenuSub,
    {
      onOpenChange: n,
      children: t
    }
  ) : /* @__PURE__ */ a.jsx(c.DropdownMenu.DropdownMenu, { onOpenChange: n, children: t });
}, ir = (r) => {
  const { children: t, sub: n, ...o } = r;
  y(o);
  const s = R(), d = Ue(
    () => ar(
      s.DropdownMenu.DropdownMenuTrigger
    ),
    [s.DropdownMenu.DropdownMenuTrigger]
  );
  return n ? /* @__PURE__ */ a.jsx(s.DropdownMenu.DropdownMenuSubTrigger, { children: t }) : /* @__PURE__ */ a.jsx(d, { asChild: !0, ...o, children: t });
}, lr = C((r, t) => {
  const { className: n, children: o, sub: s, ...d } = r;
  y(d);
  const c = R();
  return s ? /* @__PURE__ */ a.jsx(
    c.DropdownMenu.DropdownMenuSubContent,
    {
      className: n,
      ref: t,
      children: o
    }
  ) : /* @__PURE__ */ a.jsx(
    c.DropdownMenu.DropdownMenuContent,
    {
      className: n,
      ref: t,
      children: o
    }
  );
}), dr = C((r, t) => {
  const { className: n, children: o, icon: s, checked: d, subTrigger: c, onClick: v, ...f } = r;
  y(f);
  const u = R();
  return c ? /* @__PURE__ */ a.jsxs(a.Fragment, { children: [
    s,
    o
  ] }) : d !== void 0 ? /* @__PURE__ */ a.jsxs(
    u.DropdownMenu.DropdownMenuCheckboxItem,
    {
      className: b(n, "bn-gap-1"),
      ref: t,
      checked: d,
      onClick: v,
      ...f,
      children: [
        s,
        o
      ]
    }
  ) : /* @__PURE__ */ a.jsxs(
    u.DropdownMenu.DropdownMenuItem,
    {
      className: n,
      ref: t,
      onClick: v,
      ...f,
      children: [
        s,
        o,
        c && /* @__PURE__ */ a.jsx(Je, { className: "bn-ml-auto bn-h-4 bn-w-4" })
      ]
    }
  );
}), cr = C((r, t) => {
  const { className: n, ...o } = r;
  y(o);
  const s = R();
  return /* @__PURE__ */ a.jsx(
    s.DropdownMenu.DropdownMenuSeparator,
    {
      className: n,
      ref: t
    }
  );
}), ur = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  y(s);
  const d = R();
  return /* @__PURE__ */ a.jsx(
    d.DropdownMenu.DropdownMenuLabel,
    {
      className: n,
      ref: t,
      children: o
    }
  );
}), br = C((r, t) => {
  const {
    className: n,
    tabs: o,
    defaultOpenTab: s,
    openTab: d,
    setOpenTab: c,
    loading: v,
    // TODO: implement loader
    ...f
  } = r;
  y(f);
  const u = R();
  return /* @__PURE__ */ a.jsxs(
    u.Tabs.Tabs,
    {
      className: b(n, "bn-bg-popover bn-p-2 bn-rounded-lg"),
      ref: t,
      value: d,
      defaultValue: s,
      onValueChange: c,
      children: [
        /* @__PURE__ */ a.jsx(u.Tabs.TabsList, { children: o.map((g) => /* @__PURE__ */ a.jsx(u.Tabs.TabsTrigger, { value: g.name, children: g.name }, g.name)) }),
        o.map((g) => /* @__PURE__ */ a.jsx(u.Tabs.TabsContent, { value: g.name, children: /* @__PURE__ */ a.jsx(u.Card.Card, { children: /* @__PURE__ */ a.jsx(u.Card.CardContent, { className: "bn-p-4", children: g.tabPanel }) }) }, g.name))
      ]
    }
  );
}), fr = C((r, t) => {
  const { className: n, children: o, onClick: s, label: d, ...c } = r;
  y(c);
  const v = R();
  return /* @__PURE__ */ a.jsx(
    v.Button.Button,
    {
      type: "submit",
      className: n,
      "aria-label": d,
      ref: t,
      onClick: s,
      children: o
    }
  );
}), mr = C((r, t) => {
  const { className: n, accept: o, value: s, placeholder: d, onChange: c, ...v } = r;
  y(v);
  const f = R();
  return /* @__PURE__ */ a.jsx(
    f.Input.Input,
    {
      type: "file",
      className: n,
      ref: t,
      accept: o,
      value: s ? s.name : void 0,
      onChange: async (u) => c == null ? void 0 : c(u.target.files[0]),
      placeholder: d
    }
  );
}), pr = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  return y(s), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: b(
        n,
        "bn-flex bn-flex-col bn-gap-2 bn-items-start bn-justify-center"
      ),
      ref: t,
      children: o
    }
  );
}), gr = C((r, t) => {
  const { className: n, value: o, placeholder: s, onKeyDown: d, onChange: c, ...v } = r;
  y(v);
  const f = R();
  return /* @__PURE__ */ a.jsx(
    f.Input.Input,
    {
      "data-test": "embed-input",
      className: b(n, "bn-w-80"),
      ref: t,
      value: o,
      placeholder: s,
      onKeyDown: d,
      onChange: c
    }
  );
}), vr = (r) => {
  const {
    children: t,
    opened: n,
    position: o,
    // unused
    ...s
  } = r;
  y(s);
  const d = R();
  return /* @__PURE__ */ a.jsx(d.Popover.Popover, { open: n, children: t });
}, xr = C(
  (r, t) => {
    const { children: n, ...o } = r;
    y(o);
    const s = R();
    return /* @__PURE__ */ a.jsx(s.Popover.PopoverTrigger, { ref: t, asChild: !0, children: n });
  }
), hr = C((r, t) => {
  const { className: n, variant: o, children: s, ...d } = r;
  y(d);
  const c = R();
  return /* @__PURE__ */ a.jsx(
    c.Popover.PopoverContent,
    {
      sideOffset: 8,
      className: b(
        n,
        "bn-flex bn-flex-col bn-gap-2",
        o === "panel-popover" ? "bn-p-0 bn-border-none bn-shadow-none bn-max-w-none bn-w-fit" : ""
      ),
      ref: t,
      children: s
    }
  );
}), yr = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  return y(s, !1), /* @__PURE__ */ a.jsx("div", { className: n, ref: t, ...s, children: o });
}), wr = C((r, t) => {
  const {
    className: n,
    children: o,
    icon: s,
    onClick: d,
    onDragEnd: c,
    onDragStart: v,
    draggable: f,
    label: u,
    ...g
  } = r;
  y(g, !1);
  const j = R();
  return /* @__PURE__ */ a.jsxs(
    j.Button.Button,
    {
      variant: "ghost",
      className: b(n, "bn-text-gray-400"),
      ref: t,
      "aria-label": u,
      onClick: d,
      onDragStart: v,
      onDragEnd: c,
      draggable: f,
      ...g,
      children: [
        s,
        o
      ]
    }
  );
}), Cr = C((r, t) => {
  const { className: n, children: o, id: s, ...d } = r;
  return y(d), /* @__PURE__ */ a.jsx(
    "div",
    {
      id: s,
      role: "listbox",
      className: b(
        "bn-z-50 bn-min-w-[8rem] bn-overflow-auto bn-rounded-md bn-border bn-bg-popover bn-p-1 bn-text-popover-foreground bn-shadow-md data-[state=open]:bn-animate-in data-[state=closed]:bn-animate-out data-[state=closed]:bn-fade-out-0 data-[state=open]:bn-fade-in-0 data-[state=closed]:bn-zoom-out-95 data-[state=open]:bn-zoom-in-95 data-[side=bottom]:bn-slide-in-from-top-2 data-[side=left]:bn-slide-in-from-right-2 data-[side=right]:bn-slide-in-from-left-2 data-[side=top]:bn-slide-in-from-bottom-2",
        n
      ),
      ref: t,
      children: o
    }
  );
}), Nr = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  return y(s), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: b(
        "bn-relative bn-flex bn-cursor-default bn-select-none bn-items-center bn-rounded-sm bn-px-2 bn-py-1.5 bn-text-sm bn-outline-none bn-transition-colors focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
        n
      ),
      ref: t,
      children: /* @__PURE__ */ a.jsx("div", { children: o })
    }
  );
}), jr = C((r, t) => {
  const n = R(), { className: o, item: s, isSelected: d, onClick: c, id: v, ...f } = r;
  y(f);
  const u = We(null);
  return Ye(() => {
    if (!u.current || !d)
      return;
    const g = Ge(
      u.current,
      document.querySelector(".bn-suggestion-menu")
    );
    g === "top" ? u.current.scrollIntoView(!0) : g === "bottom" && u.current.scrollIntoView(!1);
  }, [d]), /* @__PURE__ */ a.jsxs(
    "div",
    {
      className: b(
        "bn-relative bn-flex bn-cursor-pointer bn-select-none bn-items-center bn-rounded-sm bn-px-2 bn-py-1.5 bn-text-sm bn-outline-none bn-transition-colors focus:bn-bg-accent focus:bn-text-accent-foreground data-[disabled]:bn-pointer-events-none data-[disabled]:bn-opacity-50",
        o
      ),
      ref: Ke([t, u]),
      id: v,
      onClick: c,
      role: "option",
      "aria-selected": d || void 0,
      children: [
        s.icon && /* @__PURE__ */ a.jsx("div", { className: "bn-p-3", "data-position": "left", children: s.icon }),
        /* @__PURE__ */ a.jsxs("div", { className: "bn-flex-1", children: [
          /* @__PURE__ */ a.jsx("div", { className: "bn-text-base", children: s.title }),
          /* @__PURE__ */ a.jsx("div", { className: "bn-text-xs", children: s.subtext })
        ] }),
        s.badge && /* @__PURE__ */ a.jsx("div", { "data-position": "right", className: "bn-text-xs", children: /* @__PURE__ */ a.jsx(n.Badge.Badge, { variant: "secondary", children: s.badge }) })
      ]
    }
  );
}), Sr = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  return y(s), /* @__PURE__ */ a.jsx(
    "div",
    {
      className: b("bn-px-2 bn-py-1.5 bn-text-sm bn-font-semibold", n),
      ref: t,
      children: o
    }
  );
}), Tr = C((r, t) => {
  const { className: n, children: o, ...s } = r;
  return y(s), /* @__PURE__ */ a.jsx("div", { className: n, ref: t, children: o });
}), Rr = C((r, t) => {
  const {
    className: n,
    children: o,
    draggable: s,
    onDragStart: d,
    onDragEnd: c,
    style: v,
    label: f,
    ...u
  } = r;
  y(u, !1);
  const g = R();
  return /* @__PURE__ */ a.jsx(
    g.Button.Button,
    {
      variant: "ghost",
      className: b(n, "bn-p-0 bn-h-fit bn-w-fit bn-text-gray-400"),
      ref: t,
      "aria-label": f,
      draggable: s,
      onDragStart: d,
      onDragEnd: c,
      style: v,
      ...u,
      children: o
    }
  );
}), Er = C((r, t) => {
  const {
    className: n,
    name: o,
    label: s,
    icon: d,
    // TODO: implement
    value: c,
    autoFocus: v,
    placeholder: f,
    onKeyDown: u,
    onChange: g,
    onSubmit: j,
    ..._
  } = r;
  y(_);
  const F = R();
  return s ? /* @__PURE__ */ a.jsxs("div", { children: [
    /* @__PURE__ */ a.jsx(F.Label.Label, { htmlFor: s, children: s }),
    /* @__PURE__ */ a.jsx(
      F.Input.Input,
      {
        className: n,
        id: s,
        name: o,
        autoFocus: v,
        placeholder: f,
        value: c,
        onKeyDown: u,
        onChange: g,
        onSubmit: j
      }
    )
  ] }) : /* @__PURE__ */ a.jsx(
    F.Input.Input,
    {
      "aria-label": o,
      name: o,
      autoFocus: v,
      placeholder: f,
      value: c,
      onKeyDown: u,
      onChange: g,
      onSubmit: j,
      ref: t
    }
  );
}), $e = C(
  (r, t) => {
    const { className: n, children: o, onMouseEnter: s, onMouseLeave: d, ...c } = r;
    y(c);
    const v = R();
    return /* @__PURE__ */ a.jsx(v.Tooltip.TooltipProvider, { delayDuration: 0, children: /* @__PURE__ */ a.jsx(
      "div",
      {
        className: b(
          n,
          "bn-flex bn-gap-1 bn-p-1 bn-bg-popover bn-text-popover-foreground bn-border bn-rounded-lg bn-shadow-md"
        ),
        ref: t,
        onMouseEnter: s,
        onMouseLeave: d,
        children: o
      }
    ) });
  }
), Ae = C(
  (r, t) => {
    const {
      className: n,
      children: o,
      mainTooltip: s,
      secondaryTooltip: d,
      icon: c,
      isSelected: v,
      isDisabled: f,
      onClick: u,
      label: g,
      ...j
    } = r;
    y(j, !1);
    const _ = R(), F = v === void 0 ? /* @__PURE__ */ a.jsxs(
      _.Button.Button,
      {
        className: n,
        variant: "ghost",
        disabled: f,
        onClick: u,
        ref: t,
        "aria-label": g,
        ...j,
        children: [
          c,
          o
        ]
      }
    ) : /* @__PURE__ */ a.jsxs(
      _.Toggle.Toggle,
      {
        className: b(
          n,
          "data-[state=open]:bg-accent data-[state=closed]:text-accent-foreground"
        ),
        "aria-label": g,
        onClick: u,
        pressed: v,
        disabled: f,
        "data-state": v ? "on" : "off",
        "data-disabled": f,
        ref: t,
        ...j,
        children: [
          c,
          o
        ]
      }
    );
    return /* @__PURE__ */ a.jsxs(_.Tooltip.Tooltip, { children: [
      /* @__PURE__ */ a.jsx(_.Tooltip.TooltipTrigger, { asChild: !0, children: F }),
      /* @__PURE__ */ a.jsxs(
        _.Tooltip.TooltipContent,
        {
          className: "bn-flex bn-flex-col bn-items-center",
          children: [
            /* @__PURE__ */ a.jsx("span", { children: s }),
            d && /* @__PURE__ */ a.jsx("span", { children: d })
          ]
        }
      )
    ] });
  }
), Dr = C((r, t) => {
  const { className: n, items: o, isDisabled: s, ...d } = r;
  y(d);
  const c = R(), v = (u) => /* @__PURE__ */ a.jsxs("div", { className: "bn-flex bn-gap-1 bn-items-center", children: [
    u.icon,
    u.text
  ] }), f = o.filter((u) => u.isSelected)[0];
  return f ? /* @__PURE__ */ a.jsxs(
    c.Select.Select,
    {
      value: f.text,
      onValueChange: (u) => {
        var g, j;
        return (j = (g = o.find((_) => _.text === u)).onClick) == null ? void 0 : j.call(g);
      },
      disabled: s,
      children: [
        /* @__PURE__ */ a.jsx(c.Select.SelectTrigger, { className: "bn-border-none", children: /* @__PURE__ */ a.jsx(c.Select.SelectValue, {}) }),
        /* @__PURE__ */ a.jsx(c.Select.SelectContent, { className: n, ref: t, children: o.map((u) => /* @__PURE__ */ a.jsx(
          c.Select.SelectItem,
          {
            disabled: u.isDisabled,
            value: u.text,
            children: /* @__PURE__ */ a.jsx(v, { ...u })
          },
          u.text
        )) })
      ]
    }
  ) : null;
}), _r = {
  FormattingToolbar: {
    Root: $e,
    Button: Ae,
    Select: Dr
  },
  FilePanel: {
    Root: br,
    Button: fr,
    FileInput: mr,
    TabPanel: pr,
    TextInput: gr
  },
  LinkToolbar: {
    Root: $e,
    Button: Ae
  },
  SideMenu: {
    Root: yr,
    Button: wr
  },
  SuggestionMenu: {
    Root: Cr,
    Item: jr,
    EmptyItem: Nr,
    Label: Sr,
    Loader: Tr
  },
  GridSuggestionMenu: {
    Root: nr,
    Item: rr,
    EmptyItem: tr,
    Loader: or
  },
  TableHandle: {
    Root: Rr
  },
  Generic: {
    Form: {
      Root: er,
      TextInput: Er
    },
    Menu: {
      Root: sr,
      Trigger: ir,
      Dropdown: lr,
      Divider: cr,
      Label: ur,
      Item: dr
    },
    Popover: {
      Root: vr,
      Trigger: xr,
      Content: hr
    }
  }
}, $r = (r) => {
  const { className: t, shadCNComponents: n, ...o } = r, s = Ue(() => ({
    ...Qt,
    ...n
  }), [n]);
  return /* @__PURE__ */ a.jsx(jn.Provider, { value: s, children: /* @__PURE__ */ a.jsx(ut.Provider, { value: _r, children: /* @__PURE__ */ a.jsx(
    bt,
    {
      className: ct("bn-shadcn", t || ""),
      ...o
    }
  ) }) });
};
export {
  $r as BlockNoteView,
  _r as components
};
