// @blocknote/ariakit/dist/blocknote-ariakit.js
import Ae, { forwardRef as y, useRef as Le, useEffect as $e } from "react";
import { assertEmpty as b, mergeCSSClasses as E, isSafari as Pr } from "@blocknote/core";
import { elementOverflow as We, mergeRefs as Ve, ComponentsContext as Cr, BlockNoteViewRaw as wr } from "@blocknote/react";
import { FormProvider as ne, FormLabel as Nr, FormInput as ae, Menu as Or, MenuButton as Ye, MenuItem as Ne, MenuButtonArrow as Ir, CheckboxCheck as Oe, MenuGroupLabel as Dr, MenuSeparator as Mr, MenuProvider as Fr, TabProvider as Ar, TabList as Lr, Tab as $r, TabPanel as Wr, Button as se, PopoverDisclosure as Vr, Popover as Yr, PopoverProvider as Br, Group as Be, Toolbar as Ur, TooltipProvider as Gr, TooltipAnchor as qr, ToolbarItem as Ue, Tooltip as Kr, SelectProvider as Jr, Select as Hr, SelectArrow as zr, SelectPopover as Xr, SelectItem as Zr, SelectItemCheck as Qr } from "@ariakit/react";
var te = { exports: {} }, W = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ie;
function et() {
  if (Ie) return W;
  Ie = 1;
  var n = Ae, a = Symbol.for("react.element"), o = Symbol.for("react.fragment"), t = Object.prototype.hasOwnProperty, l = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, u = { key: !0, ref: !0, __self: !0, __source: !0 };
  function d(g, c, v) {
    var j, _ = {}, w = null, Y = null;
    v !== void 0 && (w = "" + v), c.key !== void 0 && (w = "" + c.key), c.ref !== void 0 && (Y = c.ref);
    for (j in c) t.call(c, j) && !u.hasOwnProperty(j) && (_[j] = c[j]);
    if (g && g.defaultProps) for (j in c = g.defaultProps, c) _[j] === void 0 && (_[j] = c[j]);
    return { $$typeof: a, type: g, key: w, ref: Y, props: _, _owner: l.current };
  }
  return W.Fragment = o, W.jsx = d, W.jsxs = d, W;
}
var V = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var De;
function rt() {
  return De || (De = 1, process.env.NODE_ENV !== "production" && function() {
    var n = Ae, a = Symbol.for("react.element"), o = Symbol.for("react.portal"), t = Symbol.for("react.fragment"), l = Symbol.for("react.strict_mode"), u = Symbol.for("react.profiler"), d = Symbol.for("react.provider"), g = Symbol.for("react.context"), c = Symbol.for("react.forward_ref"), v = Symbol.for("react.suspense"), j = Symbol.for("react.suspense_list"), _ = Symbol.for("react.memo"), w = Symbol.for("react.lazy"), Y = Symbol.for("react.offscreen"), oe = Symbol.iterator, Ge = "@@iterator";
    function qe(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = oe && e[oe] || e[Ge];
      return typeof r == "function" ? r : null;
    }
    var D = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function k(e) {
      {
        for (var r = arguments.length, s = new Array(r > 1 ? r - 1 : 0), f = 1; f < r; f++)
          s[f - 1] = arguments[f];
        Ke("error", e, s);
      }
    }
    function Ke(e, r, s) {
      {
        var f = D.ReactDebugCurrentFrame, h = f.getStackAddendum();
        h !== "" && (r += "%s", s = s.concat([h]));
        var x = s.map(function(p) {
          return String(p);
        });
        x.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, x);
      }
    }
    var Je = !1, He = !1, ze = !1, Xe = !1, Ze = !1, ie;
    ie = Symbol.for("react.module.reference");
    function Qe(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === t || e === u || Ze || e === l || e === v || e === j || Xe || e === Y || Je || He || ze || typeof e == "object" && e !== null && (e.$$typeof === w || e.$$typeof === _ || e.$$typeof === d || e.$$typeof === g || e.$$typeof === c || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === ie || e.getModuleId !== void 0));
    }
    function er(e, r, s) {
      var f = e.displayName;
      if (f)
        return f;
      var h = r.displayName || r.name || "";
      return h !== "" ? s + "(" + h + ")" : s;
    }
    function le(e) {
      return e.displayName || "Context";
    }
    function N(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && k("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case t:
          return "Fragment";
        case o:
          return "Portal";
        case u:
          return "Profiler";
        case l:
          return "StrictMode";
        case v:
          return "Suspense";
        case j:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case g:
            var r = e;
            return le(r) + ".Consumer";
          case d:
            var s = e;
            return le(s._context) + ".Provider";
          case c:
            return er(e, e.render, "ForwardRef");
          case _:
            var f = e.displayName || null;
            return f !== null ? f : N(e.type) || "Memo";
          case w: {
            var h = e, x = h._payload, p = h._init;
            try {
              return N(p(x));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var O = Object.assign, A = 0, ce, ue, fe, de, me, ve, pe;
    function be() {
    }
    be.__reactDisabledLog = !0;
    function rr() {
      {
        if (A === 0) {
          ce = console.log, ue = console.info, fe = console.warn, de = console.error, me = console.group, ve = console.groupCollapsed, pe = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: be,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        A++;
      }
    }
    function tr() {
      {
        if (A--, A === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: O({}, e, {
              value: ce
            }),
            info: O({}, e, {
              value: ue
            }),
            warn: O({}, e, {
              value: fe
            }),
            error: O({}, e, {
              value: de
            }),
            group: O({}, e, {
              value: me
            }),
            groupCollapsed: O({}, e, {
              value: ve
            }),
            groupEnd: O({}, e, {
              value: pe
            })
          });
        }
        A < 0 && k("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var K = D.ReactCurrentDispatcher, J;
    function B(e, r, s) {
      {
        if (J === void 0)
          try {
            throw Error();
          } catch (h) {
            var f = h.stack.trim().match(/\n( *(at )?)/);
            J = f && f[1] || "";
          }
        return `
` + J + e;
      }
    }
    var H = !1, U;
    {
      var nr = typeof WeakMap == "function" ? WeakMap : Map;
      U = new nr();
    }
    function ge(e, r) {
      if (!e || H)
        return "";
      {
        var s = U.get(e);
        if (s !== void 0)
          return s;
      }
      var f;
      H = !0;
      var h = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var x;
      x = K.current, K.current = null, rr();
      try {
        if (r) {
          var p = function() {
            throw Error();
          };
          if (Object.defineProperty(p.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(p, []);
            } catch (P) {
              f = P;
            }
            Reflect.construct(e, [], p);
          } else {
            try {
              p.call();
            } catch (P) {
              f = P;
            }
            e.call(p.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (P) {
            f = P;
          }
          e();
        }
      } catch (P) {
        if (P && f && typeof P.stack == "string") {
          for (var m = P.stack.split(`
`), S = f.stack.split(`
`), R = m.length - 1, T = S.length - 1; R >= 1 && T >= 0 && m[R] !== S[T]; )
            T--;
          for (; R >= 1 && T >= 0; R--, T--)
            if (m[R] !== S[T]) {
              if (R !== 1 || T !== 1)
                do
                  if (R--, T--, T < 0 || m[R] !== S[T]) {
                    var C = `
` + m[R].replace(" at new ", " at ");
                    return e.displayName && C.includes("<anonymous>") && (C = C.replace("<anonymous>", e.displayName)), typeof e == "function" && U.set(e, C), C;
                  }
                while (R >= 1 && T >= 0);
              break;
            }
        }
      } finally {
        H = !1, K.current = x, tr(), Error.prepareStackTrace = h;
      }
      var F = e ? e.displayName || e.name : "", I = F ? B(F) : "";
      return typeof e == "function" && U.set(e, I), I;
    }
    function ar(e, r, s) {
      return ge(e, !1);
    }
    function sr(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function G(e, r, s) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return ge(e, sr(e));
      if (typeof e == "string")
        return B(e);
      switch (e) {
        case v:
          return B("Suspense");
        case j:
          return B("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case c:
            return ar(e.render);
          case _:
            return G(e.type, r, s);
          case w: {
            var f = e, h = f._payload, x = f._init;
            try {
              return G(x(h), r, s);
            } catch {
            }
          }
        }
      return "";
    }
    var L = Object.prototype.hasOwnProperty, he = {}, xe = D.ReactDebugCurrentFrame;
    function q(e) {
      if (e) {
        var r = e._owner, s = G(e.type, e._source, r ? r.type : null);
        xe.setExtraStackFrame(s);
      } else
        xe.setExtraStackFrame(null);
    }
    function or(e, r, s, f, h) {
      {
        var x = Function.call.bind(L);
        for (var p in e)
          if (x(e, p)) {
            var m = void 0;
            try {
              if (typeof e[p] != "function") {
                var S = Error((f || "React class") + ": " + s + " type `" + p + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[p] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw S.name = "Invariant Violation", S;
              }
              m = e[p](r, p, f, s, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (R) {
              m = R;
            }
            m && !(m instanceof Error) && (q(h), k("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", f || "React class", s, p, typeof m), q(null)), m instanceof Error && !(m.message in he) && (he[m.message] = !0, q(h), k("Failed %s type: %s", s, m.message), q(null));
          }
      }
    }
    var ir = Array.isArray;
    function z(e) {
      return ir(e);
    }
    function lr(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, s = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return s;
      }
    }
    function cr(e) {
      try {
        return ye(e), !1;
      } catch {
        return !0;
      }
    }
    function ye(e) {
      return "" + e;
    }
    function je(e) {
      if (cr(e))
        return k("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", lr(e)), ye(e);
    }
    var $ = D.ReactCurrentOwner, ur = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, Re, Ee, X;
    X = {};
    function fr(e) {
      if (L.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function dr(e) {
      if (L.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function mr(e, r) {
      if (typeof e.ref == "string" && $.current && r && $.current.stateNode !== r) {
        var s = N($.current.type);
        X[s] || (k('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', N($.current.type), e.ref), X[s] = !0);
      }
    }
    function vr(e, r) {
      {
        var s = function() {
          Re || (Re = !0, k("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        s.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: s,
          configurable: !0
        });
      }
    }
    function pr(e, r) {
      {
        var s = function() {
          Ee || (Ee = !0, k("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        s.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: s,
          configurable: !0
        });
      }
    }
    var br = function(e, r, s, f, h, x, p) {
      var m = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: a,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: s,
        props: p,
        // Record the component responsible for creating this element.
        _owner: x
      };
      return m._store = {}, Object.defineProperty(m._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(m, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: f
      }), Object.defineProperty(m, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: h
      }), Object.freeze && (Object.freeze(m.props), Object.freeze(m)), m;
    };
    function gr(e, r, s, f, h) {
      {
        var x, p = {}, m = null, S = null;
        s !== void 0 && (je(s), m = "" + s), dr(r) && (je(r.key), m = "" + r.key), fr(r) && (S = r.ref, mr(r, h));
        for (x in r)
          L.call(r, x) && !ur.hasOwnProperty(x) && (p[x] = r[x]);
        if (e && e.defaultProps) {
          var R = e.defaultProps;
          for (x in R)
            p[x] === void 0 && (p[x] = R[x]);
        }
        if (m || S) {
          var T = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          m && vr(p, T), S && pr(p, T);
        }
        return br(e, m, S, h, f, $.current, p);
      }
    }
    var Z = D.ReactCurrentOwner, Te = D.ReactDebugCurrentFrame;
    function M(e) {
      if (e) {
        var r = e._owner, s = G(e.type, e._source, r ? r.type : null);
        Te.setExtraStackFrame(s);
      } else
        Te.setExtraStackFrame(null);
    }
    var Q;
    Q = !1;
    function ee(e) {
      return typeof e == "object" && e !== null && e.$$typeof === a;
    }
    function _e() {
      {
        if (Z.current) {
          var e = N(Z.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function hr(e) {
      return "";
    }
    var ke = {};
    function xr(e) {
      {
        var r = _e();
        if (!r) {
          var s = typeof e == "string" ? e : e.displayName || e.name;
          s && (r = `

Check the top-level render call using <` + s + ">.");
        }
        return r;
      }
    }
    function Se(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var s = xr(r);
        if (ke[s])
          return;
        ke[s] = !0;
        var f = "";
        e && e._owner && e._owner !== Z.current && (f = " It was passed a child from " + N(e._owner.type) + "."), M(e), k('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', s, f), M(null);
      }
    }
    function Pe(e, r) {
      {
        if (typeof e != "object")
          return;
        if (z(e))
          for (var s = 0; s < e.length; s++) {
            var f = e[s];
            ee(f) && Se(f, r);
          }
        else if (ee(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var h = qe(e);
          if (typeof h == "function" && h !== e.entries)
            for (var x = h.call(e), p; !(p = x.next()).done; )
              ee(p.value) && Se(p.value, r);
        }
      }
    }
    function yr(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var s;
        if (typeof r == "function")
          s = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === c || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === _))
          s = r.propTypes;
        else
          return;
        if (s) {
          var f = N(r);
          or(s, e.props, "prop", f, e);
        } else if (r.PropTypes !== void 0 && !Q) {
          Q = !0;
          var h = N(r);
          k("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", h || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && k("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function jr(e) {
      {
        for (var r = Object.keys(e.props), s = 0; s < r.length; s++) {
          var f = r[s];
          if (f !== "children" && f !== "key") {
            M(e), k("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", f), M(null);
            break;
          }
        }
        e.ref !== null && (M(e), k("Invalid attribute `ref` supplied to `React.Fragment`."), M(null));
      }
    }
    var Ce = {};
    function we(e, r, s, f, h, x) {
      {
        var p = Qe(e);
        if (!p) {
          var m = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (m += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var S = hr();
          S ? m += S : m += _e();
          var R;
          e === null ? R = "null" : z(e) ? R = "array" : e !== void 0 && e.$$typeof === a ? (R = "<" + (N(e.type) || "Unknown") + " />", m = " Did you accidentally export a JSX literal instead of a component?") : R = typeof e, k("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", R, m);
        }
        var T = gr(e, r, s, h, x);
        if (T == null)
          return T;
        if (p) {
          var C = r.children;
          if (C !== void 0)
            if (f)
              if (z(C)) {
                for (var F = 0; F < C.length; F++)
                  Pe(C[F], e);
                Object.freeze && Object.freeze(C);
              } else
                k("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Pe(C, e);
        }
        if (L.call(r, "key")) {
          var I = N(e), P = Object.keys(r).filter(function(Sr) {
            return Sr !== "key";
          }), re = P.length > 0 ? "{key: someKey, " + P.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!Ce[I + re]) {
            var kr = P.length > 0 ? "{" + P.join(": ..., ") + ": ...}" : "{}";
            k(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, re, I, kr, I), Ce[I + re] = !0;
          }
        }
        return e === t ? jr(T) : yr(T), T;
      }
    }
    function Rr(e, r, s) {
      return we(e, r, s, !0);
    }
    function Er(e, r, s) {
      return we(e, r, s, !1);
    }
    var Tr = Er, _r = Rr;
    V.Fragment = t, V.jsx = Tr, V.jsxs = _r;
  }()), V;
}
process.env.NODE_ENV === "production" ? te.exports = et() : te.exports = rt();
var i = te.exports;
const tt = (n) => {
  const { children: a, ...o } = n;
  return b(o), /* @__PURE__ */ i.jsx(ne, { children: a });
}, nt = y((n, a) => {
  const { className: o, children: t, id: l, columns: u, ...d } = n;
  return b(d), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: o,
      style: { gridTemplateColumns: `repeat(${u}, 1fr)` },
      ref: a,
      id: l,
      role: "grid",
      children: t
    }
  );
}), at = y((n, a) => {
  const { className: o, children: t, columns: l, ...u } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: E("bn-ak-menu-item", o || ""),
      style: { gridColumn: `1 / ${l + 1}` },
      ref: a,
      children: /* @__PURE__ */ i.jsx("div", { className: "bn-ak-suggestion-menu-item-label", children: t })
    }
  );
}), st = y((n, a) => {
  const { className: o, isSelected: t, onClick: l, item: u, id: d, ...g } = n;
  b(g);
  const c = Le(null);
  return $e(() => {
    if (!c.current || !t)
      return;
    const v = We(
      c.current,
      document.querySelector(".bn-grid-suggestion-menu")
    );
    v === "top" ? c.current.scrollIntoView(!0) : v === "bottom" && c.current.scrollIntoView(!1);
  }, [t]), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: o,
      ref: Ve([a, c]),
      id: d,
      role: "option",
      onClick: l,
      "aria-selected": t || void 0,
      children: u.icon
    }
  );
}), ot = y((n, a) => {
  const {
    className: o,
    children: t,
    // unused, using "dots" instead
    columns: l,
    ...u
  } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: o,
      style: { gridColumn: `1 / ${l + 1}` },
      ref: a,
      children: t
    }
  );
}), it = y((n, a) => {
  const {
    className: o,
    name: t,
    label: l,
    icon: u,
    value: d,
    autoFocus: g,
    placeholder: c,
    onKeyDown: v,
    onChange: j,
    onSubmit: _,
    ...w
  } = n;
  return b(w), /* @__PURE__ */ i.jsxs(i.Fragment, { children: [
    n.label && /* @__PURE__ */ i.jsx(Nr, { name: t, children: l }),
    /* @__PURE__ */ i.jsxs("div", { className: "bn-ak-input-wrapper", children: [
      u,
      /* @__PURE__ */ i.jsx(
        ae,
        {
          className: E("bn-ak-input", o || ""),
          ref: a,
          name: t,
          value: d,
          autoFocus: g,
          placeholder: c,
          onKeyDown: v,
          onChange: j,
          onSubmit: _
        }
      )
    ] })
  ] });
}), lt = (n) => {
  const {
    children: a,
    onOpenChange: o,
    position: t,
    sub: l,
    // unused
    ...u
  } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    Fr,
    {
      placement: t,
      setOpen: o,
      virtualFocus: !0,
      children: a
    }
  );
}, ct = y((n, a) => {
  const {
    className: o,
    children: t,
    sub: l,
    // unused
    ...u
  } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    Or,
    {
      unmountOnHide: !0,
      className: E("bn-ak-menu", o || ""),
      ref: a,
      children: t
    }
  );
}), ut = y((n, a) => {
  const { className: o, children: t, icon: l, checked: u, subTrigger: d, onClick: g, ...c } = n;
  return b(c), d ? /* @__PURE__ */ i.jsxs(
    Ye,
    {
      render: /* @__PURE__ */ i.jsx(Ne, {}),
      className: E("bn-ak-menu-item", o || ""),
      ref: a,
      onClick: g,
      children: [
        l,
        t,
        /* @__PURE__ */ i.jsx(Ir, {}),
        u !== void 0 && /* @__PURE__ */ i.jsx(Oe, { checked: u })
      ]
    }
  ) : /* @__PURE__ */ i.jsxs(
    Ne,
    {
      className: E("bn-ak-menu-item", o || ""),
      ref: a,
      onClick: g,
      children: [
        l,
        t,
        u !== void 0 && /* @__PURE__ */ i.jsx(Oe, { checked: u })
      ]
    }
  );
}), ft = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx(
    Dr,
    {
      className: E("bn-ak-group-label", o || ""),
      ref: a,
      children: t
    }
  );
}), dt = (n) => {
  const { children: a, sub: o, ...t } = n;
  return b(t), o ? a : /* @__PURE__ */ i.jsx(Ye, { render: a });
}, mt = y((n, a) => {
  const { className: o, ...t } = n;
  return b(t), /* @__PURE__ */ i.jsx(
    Mr,
    {
      className: E("bn-ak-separator", o || ""),
      ref: a
    }
  );
}), vt = y((n, a) => {
  const {
    className: o,
    tabs: t,
    defaultOpenTab: l,
    openTab: u,
    setOpenTab: d,
    loading: g,
    // TODO: implement loading
    ...c
  } = n;
  return b(c), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: E("bn-ak-wrapper", o || ""),
      ref: a,
      children: /* @__PURE__ */ i.jsxs(
        Ar,
        {
          defaultSelectedId: l,
          selectedId: u,
          setActiveId: (v) => {
            v && d(v);
          },
          children: [
            /* @__PURE__ */ i.jsx(Lr, { className: "bn-ak-tab-list", children: t.map((v) => /* @__PURE__ */ i.jsx($r, { className: "bn-ak-tab", id: v.name, children: v.name }, v.name)) }),
            /* @__PURE__ */ i.jsx("div", { className: "bn-ak-panels", children: t.map((v) => /* @__PURE__ */ i.jsx(Wr, { tabId: v.name, children: v.tabPanel }, v.name)) })
          ]
        }
      )
    }
  );
}), pt = y((n, a) => {
  const { className: o, children: t, onClick: l, label: u, ...d } = n;
  return b(d), /* @__PURE__ */ i.jsx(
    se,
    {
      className: E("bn-ak-button", o || ""),
      onClick: l,
      "aria-label": u,
      ref: a,
      children: t
    }
  );
}), bt = y((n, a) => {
  const { className: o, accept: t, value: l, placeholder: u, onChange: d, ...g } = n;
  return b(g), /* @__PURE__ */ i.jsx(ne, { children: /* @__PURE__ */ i.jsx(
    ae,
    {
      className: o,
      ref: a,
      name: "panel-input",
      type: "file",
      accept: t,
      value: l ? l.name : void 0,
      onChange: async (c) => d == null ? void 0 : d(c.target.files[0]),
      placeholder: u
    }
  ) });
}), gt = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx("div", { className: o, ref: a, children: t });
}), ht = y((n, a) => {
  const { className: o, value: t, placeholder: l, onKeyDown: u, onChange: d, ...g } = n;
  return b(g), /* @__PURE__ */ i.jsx(ne, { children: /* @__PURE__ */ i.jsx(
    ae,
    {
      className: E("bn-ak-input", o || ""),
      name: "panel-input",
      value: t,
      placeholder: l,
      onChange: d,
      onKeyDown: u,
      "data-test": "embed-input",
      ref: a
    }
  ) });
}), xt = y((n, a) => {
  const { children: o, ...t } = n;
  return b(t), /* @__PURE__ */ i.jsx(Vr, { render: o, ref: a });
}), yt = y((n, a) => {
  const { className: o, children: t, variant: l, ...u } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    Yr,
    {
      className: E("bn-ak-popover", o || ""),
      ref: a,
      children: t
    }
  );
}), jt = (n) => {
  const { children: a, opened: o, position: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx(Br, { open: o, placement: t, children: a });
}, Rt = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l, !1), /* @__PURE__ */ i.jsx(Be, { className: o, ref: a, ...l, children: t });
}), Et = y((n, a) => {
  const {
    className: o,
    children: t,
    icon: l,
    onClick: u,
    label: d,
    onDragEnd: g,
    onDragStart: c,
    draggable: v,
    ...j
  } = n;
  return b(j, !1), /* @__PURE__ */ i.jsxs(
    se,
    {
      onDragEnd: g,
      onDragStart: c,
      draggable: v,
      "aria-label": d,
      className: E(
        "bn-ak-button bn-ak-secondary",
        o || ""
      ),
      ref: a,
      onClick: u,
      ...j,
      children: [
        l,
        t
      ]
    }
  );
}), Tt = y((n, a) => {
  const { className: o, children: t, id: l, ...u } = n;
  return b(u), /* @__PURE__ */ i.jsx(
    Be,
    {
      className: E("bn-ak-menu", o || ""),
      id: l,
      role: "listbox",
      ref: a,
      children: t
    }
  );
}), _t = y((n, a) => {
  const { className: o, item: t, isSelected: l, onClick: u, id: d, ...g } = n;
  b(g);
  const c = Le(null);
  return $e(() => {
    if (!c.current || !l)
      return;
    const v = We(
      c.current,
      document.querySelector(".bn-suggestion-menu")
    );
    v === "top" ? c.current.scrollIntoView(!0) : v === "bottom" && c.current.scrollIntoView(!1);
  }, [l]), /* @__PURE__ */ i.jsxs(
    "div",
    {
      className: E("bn-ak-menu-item", o || ""),
      ref: Ve([a, c]),
      id: d,
      onClick: u,
      role: "option",
      "aria-selected": l || void 0,
      children: [
        t.icon && /* @__PURE__ */ i.jsx(
          "div",
          {
            className: "bn-ak-suggestion-menu-item-section",
            "data-position": "left",
            children: t.icon
          }
        ),
        /* @__PURE__ */ i.jsxs("div", { className: "bn-ak-suggestion-menu-item-body", children: [
          /* @__PURE__ */ i.jsx("div", { className: "bn-ak-suggestion-menu-item-title", children: t.title }),
          /* @__PURE__ */ i.jsx("div", { className: "bn-ak-suggestion-menu-item-subtitle", children: t.subtext })
        ] }),
        t.badge && /* @__PURE__ */ i.jsx(
          "div",
          {
            "data-position": "right",
            className: "bn-ak-suggestion-menu-item-section",
            children: /* @__PURE__ */ i.jsx("div", { children: t.badge })
          }
        )
      ]
    }
  );
}), kt = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: E("bn-ak-menu-item", o || ""),
      ref: a,
      children: /* @__PURE__ */ i.jsx("div", { className: "bn-ak-suggestion-menu-item-label", children: t })
    }
  );
}), St = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx(
    "div",
    {
      className: E("bn-ak-group-label", o || ""),
      ref: a,
      children: t
    }
  );
}), Pt = y((n, a) => {
  const { className: o, children: t, ...l } = n;
  return b(l), /* @__PURE__ */ i.jsx("div", { className: o, ref: a, children: t });
}), Ct = y((n, a) => {
  const {
    className: o,
    children: t,
    draggable: l,
    onDragStart: u,
    onDragEnd: d,
    style: g,
    label: c,
    ...v
  } = n;
  return b(v, !1), /* @__PURE__ */ i.jsx(
    se,
    {
      className: E(
        "bn-ak-button bn-ak-secondary",
        o || ""
      ),
      ref: a,
      "aria-label": c,
      draggable: l,
      onDragStart: u,
      onDragEnd: d,
      style: g,
      ...v,
      children: t
    }
  );
}), Me = y(
  (n, a) => {
    const { className: o, children: t, onMouseEnter: l, onMouseLeave: u, ...d } = n;
    return b(d), /* @__PURE__ */ i.jsx(
      Ur,
      {
        className: E("bn-ak-toolbar", o || ""),
        ref: a,
        onMouseEnter: l,
        onMouseLeave: u,
        children: t
      }
    );
  }
), Fe = y(
  (n, a) => {
    const {
      className: o,
      children: t,
      mainTooltip: l,
      secondaryTooltip: u,
      icon: d,
      isSelected: g,
      isDisabled: c,
      onClick: v,
      label: j,
      ..._
    } = n;
    return b(_, !1), /* @__PURE__ */ i.jsxs(Gr, { children: [
      /* @__PURE__ */ i.jsx(
        qr,
        {
          className: "link",
          render: /* @__PURE__ */ i.jsxs(
            Ue,
            {
              "aria-label": j,
              className: E(
                "bn-ak-button bn-ak-secondary",
                o || ""
              ),
              onMouseDown: (w) => {
                Pr() && w.currentTarget.focus();
              },
              onClick: v,
              "aria-pressed": g,
              "data-selected": g ? "true" : void 0,
              "data-test": n.mainTooltip.slice(0, 1).toLowerCase() + n.mainTooltip.replace(/\s+/g, "").slice(1),
              disabled: c || !1,
              ref: a,
              ..._,
              children: [
                d,
                t
              ]
            }
          )
        }
      ),
      /* @__PURE__ */ i.jsxs(Kr, { className: "bn-ak-tooltip", children: [
        /* @__PURE__ */ i.jsx("span", { children: l }),
        u && /* @__PURE__ */ i.jsx("span", { children: u })
      ] })
    ] });
  }
), wt = y((n, a) => {
  const { className: o, items: t, isDisabled: l, ...u } = n;
  b(u);
  const d = n.items.filter((c) => c.isSelected)[0], g = (c) => {
    var v, j;
    (j = (v = t.find((_) => _.text === c)).onClick) == null || j.call(v);
  };
  return /* @__PURE__ */ i.jsxs(Jr, { value: d.text, setValue: g, children: [
    /* @__PURE__ */ i.jsxs(
      Hr,
      {
        className: "bn-ak-button bn-ak-secondary",
        disabled: l,
        "aria-label": "Text alignment",
        render: /* @__PURE__ */ i.jsx(Ue, {}),
        children: [
          d.icon,
          " ",
          d.text,
          " ",
          /* @__PURE__ */ i.jsx(zr, {})
        ]
      }
    ),
    /* @__PURE__ */ i.jsx(
      Xr,
      {
        className: E("bn-ak-popover", o || ""),
        ref: a,
        gutter: 4,
        children: t.map((c) => /* @__PURE__ */ i.jsxs(
          Zr,
          {
            className: "bn-ak-select-item",
            value: c.text,
            children: [
              c.icon,
              c.text,
              c.text === d.text && /* @__PURE__ */ i.jsx(Qr, {})
            ]
          },
          c.text
        ))
      }
    )
  ] });
}), Nt = {
  FormattingToolbar: {
    Root: Me,
    Button: Fe,
    Select: wt
  },
  FilePanel: {
    Root: vt,
    Button: pt,
    FileInput: bt,
    TabPanel: gt,
    TextInput: ht
  },
  GridSuggestionMenu: {
    Root: nt,
    Item: st,
    EmptyItem: at,
    Loader: ot
  },
  LinkToolbar: {
    Root: Me,
    Button: Fe
  },
  SideMenu: {
    Root: Rt,
    Button: Et
  },
  SuggestionMenu: {
    Root: Tt,
    Item: _t,
    EmptyItem: kt,
    Label: St,
    Loader: Pt
  },
  TableHandle: {
    Root: Ct
  },
  Generic: {
    Form: {
      Root: tt,
      TextInput: it
    },
    Menu: {
      Root: lt,
      Trigger: dt,
      Dropdown: ct,
      Divider: mt,
      Label: ft,
      Item: ut
    },
    Popover: {
      Root: jt,
      Trigger: xt,
      Content: yt
    }
  }
}, Ft = (n) => {
  const { className: a, ...o } = n;
  return /* @__PURE__ */ i.jsx(Cr.Provider, { value: Nt, children: /* @__PURE__ */ i.jsx(
    wr,
    {
      className: E("bn-ariakit", a || ""),
      ...o
    }
  ) });
};
export {
  Ft as BlockNoteView,
  Nt as components
};
